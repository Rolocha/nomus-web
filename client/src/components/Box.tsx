import styled from '@emotion/styled'

import {
  space,
  SpaceProps,
  position,
  PositionProps,
  color,
  ColorProps,
  border,
  BorderProps,
  layout,
  LayoutProps,
  flexbox,
  FlexboxProps,
  grid,
  GridProps,
  shadow,
  ShadowProps,
} from 'styled-system'

type BoxProps = { as?: string } & SpaceProps &
  PositionProps &
  ColorProps &
  BorderProps &
  LayoutProps &
  FlexboxProps &
  GridProps &
  ShadowProps

const Box = styled<'div', BoxProps>('div')(
  {
    boxSizing: 'border-box',
    minWidth: 0,
  },
  space,
  position,
  color,
  border,
  layout,
  flexbox,
  grid,
  shadow,
)

export default Box
