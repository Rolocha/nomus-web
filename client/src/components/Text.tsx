import styled from '@emotion/styled'
import * as React from 'react'
import theme from 'src/styles/theme'
import { baseTextStyles } from 'src/styles/typography'
import {
  color,
  ColorProps,
  fontFamily,
  FontFamilyProps,
  fontSize,
  FontSizeProps,
  fontWeight,
  FontWeightProps,
  lineHeight,
  LineHeightProps,
  space,
  SpaceProps,
  variant,
} from 'styled-system'

export type TextProps = {
  variant?: keyof typeof theme.textStyles | null
  as?: string
} & FontSizeProps &
  ColorProps &
  FontFamilyProps &
  LineHeightProps &
  FontWeightProps &
  SpaceProps

const Text = styled<'p', TextProps>('p')(
  baseTextStyles,
  variant({
    variants: theme.textStyles,
  }),
  fontSize,
  color,
  space,
  fontFamily,
  lineHeight,
  fontWeight,
)

Text.defaultProps = {
  variant: 'body',
}

const semanticallyStyledText = (
  as: string,
  variant?: keyof typeof theme.textStyles | null,
) => (props: React.ComponentProps<typeof Text>) => (
  <Text as={as} variant={variant} {...props}>
    {props.children}
  </Text>
)

// Custom text variants ready to be used with 'as' baked in
export const Heading = semanticallyStyledText('h1', 'heading')
export const PageHeader = semanticallyStyledText('h2', 'pageHeader')
export const SectionHeader = semanticallyStyledText('h3', 'sectionHeader')
export const SectionSubheader = semanticallyStyledText('h4', 'sectionSubheader')

export const Body = semanticallyStyledText('p', 'body')
export const Body3 = semanticallyStyledText('p', 'bodyTertiary')
export const Caption = semanticallyStyledText('p', 'caption')
export const Label = semanticallyStyledText('p', 'label')
export const Plain = semanticallyStyledText('p', null)

export const Link = styled(Text.withComponent('a'))`
  cursor: pointer;
`
Link.defaultProps = {
  ...theme.textStyles.link,
}
