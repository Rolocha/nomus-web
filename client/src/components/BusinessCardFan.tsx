import * as React from 'react'
import Box from 'src/components/Box'
import Image from 'src/components/Image'
import { getImageDimensions, ImageDimensions } from 'src/utils/image'

interface Props {
  frontImageUrl: string
  backImageUrl: string
}

const BusinessCardFan = ({ frontImageUrl, backImageUrl }: Props) => {
  const [businessCardDimensions, setBusinessCardDimensions] = React.useState<
    ImageDimensions | 'determining' | null
  >(null)
  React.useEffect(() => {
    if (businessCardDimensions == null) {
      setBusinessCardDimensions('determining')
      getImageDimensions(frontImageUrl).then(setBusinessCardDimensions)
    }
  }, [frontImageUrl, businessCardDimensions])

  const cardOrientation = React.useMemo(() => {
    if (
      businessCardDimensions == null ||
      businessCardDimensions === 'determining'
    ) {
      return 'unknown'
    }
    return businessCardDimensions.height > businessCardDimensions.width
      ? 'vertical'
      : 'horizontal'
  }, [businessCardDimensions])

  return (
    <Box
      position="relative"
      visibility={cardOrientation === 'unknown' ? 'hidden' : 'visible'}
      pt={
        { vertical: '10%', horizontal: '21%', unknown: '0%' }[cardOrientation]
      }
    >
      {frontImageUrl && (
        <Image
          zIndex={2}
          position="relative"
          width={
            { horizontal: '80%', vertical: '50%', unknown: '50%' }[
              cardOrientation
            ]
          }
          ml="auto"
          src={frontImageUrl}
          boxShadow="businessCard"
        />
      )}
      {backImageUrl && (
        <Image
          zIndex={1}
          position="absolute"
          transform="rotateZ(-12.67deg)"
          {...{
            vertical: {
              width: '50%',
              bottom: '5%',
              right: '35%',
            },
            horizontal: {
              width: '80%',
              right: '15%',
              bottom: '18%',
            },
            unknown: {},
          }[cardOrientation]}
          src={backImageUrl}
          boxShadow="businessCard"
        />
      )}
    </Box>
  )
}

export default BusinessCardFan
