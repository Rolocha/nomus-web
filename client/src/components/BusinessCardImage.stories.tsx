import React from 'react'
import Box from 'src/components/Box'
import BusinessCardImage from 'src/components/BusinessCardImage'

export default {
  title: 'components/BusinessCardImage',
  component: BusinessCardImage,
  excludeStories: /.*Data$/,
}

export const Primary = () => {
  return (
    <Box>
      <BusinessCardImage
        frontImageUrl="https://placehold.it/500x300"
        backImageUrl="https://placehold.it/550x330"
        width="300px"
      />
    </Box>
  )
}
