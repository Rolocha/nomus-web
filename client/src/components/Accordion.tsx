import * as React from 'react'
import { colors } from 'src/styles'
import Box from './Box'
import * as Text from './Text'
import Icon from './Icon'

interface Props {
  title: string
  children: React.ReactNode
  size?: 'medium' | 'small'
}

const Accordion = ({ title, children, size = 'medium' }: Props) => {
  const [isOpen, setIsOpen] = React.useState(false)

  const handleButtonClick = () => {
    setIsOpen(!isOpen)
  }

  return (
    <Box>
      <Box
        borderTop={`1px solid ${colors.africanElephant}`}
        borderBottom={
          !isOpen ? `1px solid ${colors.africanElephant}` : undefined
        }
        height={
          {
            medium: '64px',
            small: '42px',
          }[size]
        }
        display="flex"
        alignItems="center"
        justifyContent="space-between"
      >
        <Text.SectionSubheader>{title}</Text.SectionSubheader>
        <Box role="button" cursor="pointer" onClick={handleButtonClick}>
          {isOpen ? (
            <Icon of="minus" color={colors.midnightGray} />
          ) : (
            <Icon of="plus" color={colors.midnightGray} />
          )}
        </Box>
      </Box>

      {isOpen && (
        <Box borderBottom={`1px solid ${colors.africanElephant}`} pb="20px">
          {children}
        </Box>
      )}
    </Box>
  )
}

export default Accordion
