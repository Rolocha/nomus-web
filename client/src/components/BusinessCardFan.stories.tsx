import React from 'react'
import Box from 'src/components/Box'
import BusinessCardFan from 'src/components/BusinessCardFan'

export default {
  title: 'components/BusinessCardFan',
  component: BusinessCardFan,
  excludeStories: /.*Data$/,
}

export const Primary = () => {
  return (
    <Box
      display="grid"
      gridTemplateColumns="400px 400px 400px"
      gridColumnGap="16px"
    >
      <BusinessCardFan
        frontImageUrl="https://nomus-assets.s3.amazonaws.com/card-versions/cardv_6153ef94d2b56002265e6a84/front/1632890782277.png"
        backImageUrl="https://nomus-assets.s3.amazonaws.com/card-versions/cardv_6153ef94d2b56002265e6a84/back/1632890783439.png"
      />
      <BusinessCardFan
        frontImageUrl="https://nomus-assets.s3.amazonaws.com/card-versions/cardv_60ae1754821500001c8f7683/front"
        backImageUrl="https://nomus-assets.s3.amazonaws.com/card-versions/cardv_60ae1754821500001c8f7683/back"
      />
      <BusinessCardFan
        frontImageUrl={null}
        backImageUrl={null}
        showPlaceholderIfMissing
      />
    </Box>
  )
}
