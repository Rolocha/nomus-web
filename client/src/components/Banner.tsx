import { css } from '@emotion/react'
import * as React from 'react'
import * as Text from 'src/components/Text'
import * as SVG from 'src/components/SVG'
import { colors } from 'src/styles'
import Box from './Box'

interface Props {
  type: 'success' | 'warning' | 'danger' | 'info'
  title: string
  description: React.ReactNode
  className?: string
  closable?: boolean
  onClickClose?: () => void
}

const Banner = ({
  type,
  title,
  className,
  description,
  closable,
  onClickClose,
}: Props) => {
  const color = {
    success: colors.validGreen,
    warning: colors.poppy,
    danger: colors.invalidRed,
    info: colors.nomusBlue,
  }[type]

  const Icon = {
    success: SVG.CheckO,
    warning: SVG.ExclamationO,
    danger: SVG.SlashO,
    info: SVG.InfoO,
  }[type]

  return (
    <Box
      className={className}
      position="relative"
      borderRadius={2}
      display="flex"
      alignItems="center"
      boxShadow="banner"
      bg={colors.white}
    >
      <Box
        bg={color}
        width="8px"
        alignSelf="stretch"
        borderTopLeftRadius={2}
        borderBottomLeftRadius={2}
      />
      <Box
        display="grid"
        gridTemplateColumns={closable ? 'auto auto' : 'auto'}
        gridColumnGap={2}
        width="100%"
        justifyContent="space-between"
        alignItems="center"
        p={2}
      >
        <Box display="flex">
          <Icon color={color} css={css({ marginRight: '8px' })} />
          <Text.Body2>
            <span css={css({ fontWeight: 500, marginRight: '8px' })}>
              {title}
            </span>
            {description}
          </Text.Body2>
        </Box>
        {closable && (
          <Box
            cursor="pointer"
            role="button"
            aria-label="close banner"
            onClick={onClickClose}
          >
            <SVG.Close color={colors.midnightGray} />
          </Box>
        )}
      </Box>
    </Box>
  )
}

export default Banner
