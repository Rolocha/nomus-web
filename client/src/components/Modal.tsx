import { css, Global } from '@emotion/core'
import * as CSS from 'csstype'
import { rgba } from 'polished'
import * as React from 'react'
import Box from 'src/components/Box'
import Button from 'src/components/Button'
import * as Text from 'src/components/Text'
import { colors } from 'src/styles'
import {
  RequiredTheme,
  ResponsiveValue,
  TLengthStyledSystem,
} from 'styled-system'

interface Action {
  text: string
  handler: () => void
}

type Actions = {
  primary?: Action
  secondary?: Action
}

interface Props {
  isOpen: boolean
  onClose: (confirmed?: boolean) => void
  // If omitted, defaults to using onClose
  onClickOutside?: () => void
  children: React.ReactNode
  width?: ResponsiveValue<
    CSS.MaxWidthProperty<TLengthStyledSystem>,
    RequiredTheme
  >
  height?: ResponsiveValue<
    CSS.MaxHeightProperty<TLengthStyledSystem>,
    RequiredTheme
  >
  confirmClose?: () => boolean
  actions?: Actions
  allowCloseWithOutsideClick?: boolean
}

const Modal = ({
  isOpen,
  children,
  onClose,
  width,
  height,
  confirmClose = () => false,
  actions,
  allowCloseWithOutsideClick = true,
}: Props) => {
  const lastIsOpen = React.useRef(isOpen)
  const modalCardRef = React.useRef<HTMLDivElement>(null)
  const [confirmingClose, setConfirmingClose] = React.useState(false)

  const handleOutsideClick = (event: React.SyntheticEvent<any>) => {
    if (
      allowCloseWithOutsideClick &&
      !confirmingClose &&
      modalCardRef.current &&
      // @ts-ignore
      !modalCardRef.current.contains(event.target)
    ) {
      tryToClose()
    }
  }

  const tryToClose = React.useCallback(() => {
    if (confirmClose()) {
      if (confirmingClose) {
        onClose(true)
        setConfirmingClose(false)
      } else {
        setConfirmingClose(true)
      }
    } else {
      onClose()
    }
  }, [confirmClose, confirmingClose, setConfirmingClose, onClose])

  const cancelCloseConfirm = React.useCallback(() => {
    setConfirmingClose(false)
  }, [setConfirmingClose])

  // Disable scroll when modal gets opened
  React.useEffect(() => {
    if (isOpen !== lastIsOpen.current) {
      if (isOpen) {
        document.querySelector('html')?.classList.add('scroll-lock')
      } else {
        document.querySelector('html')?.classList.remove('scroll-lock')
      }
      lastIsOpen.current = isOpen
    }
  }, [isOpen, lastIsOpen])

  return isOpen ? (
    <Box
      as="aside"
      aria-modal
      position="fixed"
      top="0"
      left="0"
      zIndex={10}
      width="100vw"
      height="100vh"
      display="flex"
      flexDirection="column"
      alignItems="center"
      justifyContent="center"
      bg={rgba(colors.primaryTeal, 0.8)}
      onClick={handleOutsideClick}
    >
      <Global
        styles={css`
          .scroll-lock {
            overflow: hidden;
          }
        `}
      />
      <Box
        ref={modalCardRef}
        position="relative"
        zIndex={20}
        maxWidth={width ?? { _: '95%', md: '80%' }}
        maxHeight={height ?? { _: '95%', md: '80%' }}
        minWidth="calc(min(80vw, 400px))"
        bg="white"
        borderRadius={2}
        boxShadow={0}
        display="flex"
        flexDirection="column"
        alignItems="stretch"
        justifyContent={actions ? 'space-between' : 'flex-start'}
      >
        <Box p={{ _: 3, md: 4 }} overflowX="auto">
          {children}
          <Box
            role="button"
            aria-label="Close Modal"
            onClick={tryToClose}
            p="inherit"
            position="absolute"
            top="0"
            right="0"
            css={css`
              cursor: pointer;
            `}
          >
            <Text.Plain fontSize={24}>X</Text.Plain>
          </Box>
        </Box>

        {actions && (
          <Box
            display="flex"
            flexDirection="row"
            justifyContent="flex-end"
            borderTop={`1px solid ${colors.superlightGray}`}
            p="inherit"
          >
            <Box p={3}>
              {actions.secondary && (
                <Button
                  variant="secondaryOutline"
                  mr={2}
                  onClick={actions.secondary.handler}
                >
                  {actions.secondary.text}
                </Button>
              )}
              {actions.primary && (
                <Button variant="primary" onClick={actions.primary.handler}>
                  {actions.primary.text}
                </Button>
              )}
            </Box>
          </Box>
        )}
      </Box>
      {confirmClose() && (
        <Box zIndex={30}>
          <Modal
            allowCloseWithOutsideClick={false}
            isOpen={confirmingClose}
            onClose={cancelCloseConfirm}
            actions={{
              primary: {
                text: 'No thanks',
                handler: cancelCloseConfirm,
              },
              secondary: {
                text: 'Discard',
                handler: tryToClose,
              },
            }}
          >
            <Box>
              <Box>
                <Text.PageHeader mb={3}>Discard?</Text.PageHeader>

                <Text.Body>Changes you have made won't be saved.</Text.Body>
              </Box>
            </Box>
          </Modal>
        </Box>
      )}
    </Box>
  ) : null
}

export default Modal
