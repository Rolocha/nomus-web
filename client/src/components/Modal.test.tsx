import * as React from 'react'
import renderer from 'react-test-renderer'
import { render, cleanup } from '@testing-library/react'

import { PageHeader, Body } from 'src/components/Text'
import Modal from 'src/components/Modal'

afterEach(cleanup)

describe('<Modal />', () => {
  it('renders a modal with the specified content', () => {
    const component = renderer.create(
      <Modal isOpen={true} onClose={() => {}}>
        <PageHeader>Test Modal</PageHeader>
        <Body>Some content</Body>
      </Modal>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('if confirmClose is true, show confirmation modal when trying to close', () => {
    const closeModal = jest.fn()

    const { getByText } = render(
      <Modal isOpen={true} onClose={closeModal} confirmClose={() => true}>
        <PageHeader>Test Modal</PageHeader>
        <Body>Some content</Body>
      </Modal>,
    )

    getByText('X').click()
    expect(getByText('Discard?')).toBeDefined()

    // Click discard button
    getByText((content, element) => {
      return content === 'Discard' && element.tagName.toLowerCase() === 'button'
    }).click()

    expect(closeModal).toHaveBeenCalledTimes(1)
  })
})
