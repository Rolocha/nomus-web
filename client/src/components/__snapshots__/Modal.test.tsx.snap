// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> renders a modal with the specified content 1`] = `
<aside
  aria-modal={true}
  className="css-16lu7ni-Box e1gi27mm0"
  display="flex"
  height="100vh"
  onClick={[Function]}
  width="100vw"
>
  <div
    className="css-19yvz14-Box e1gi27mm0"
    display="flex"
  >
    <div
      className="css-q4rbdj-Box e1gi27mm0"
    >
      <h2
        className="css-10tfo6t-Text e1pb1ozb0"
      >
        Test Modal
      </h2>
      <p
        className="css-h555zl-Text e1pb1ozb0"
      >
        Some content
      </p>
      <div
        aria-label="Close Modal"
        className="css-1qehuvj-Box-Modal e1gi27mm0"
        onClick={[Function]}
        role="button"
      >
        <p
          className="css-jvuolr-Text e1pb1ozb0"
          fontSize={24}
        >
          X
        </p>
      </div>
    </div>
  </div>
</aside>
`;
